### PR TITLE
proposed fix for gemm thread safety; using thread-local storage

### PR DESCRIPTION
**Pull request by guacamoleo**
each thread will have its own kernel map and, therefore, its own set of kernels

[Review on Reviewable](https://reviewable.io/reviews/clmathlibraries/clblas/235)

### CODE REVIEW

**Comment by kknox** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/library/blas/xgemm.cc, line 152 \[r1\]](https://reviewable.io:443/reviews/clmathlibraries/clblas/235#-KBszTX9eCAtl7LRXCYC:-KBszTXAF0lHhzqKn60N:318809294) ([raw file](https://github.com/clmathlibraries/clblas/blob/2b56167fbc887acaf42aa7707ab8cfcd3a8c468e/src/library/blas/xgemm.cc#L152)):*
This is a redundant #include; duplicate from top of file

---

*[src/library/blas/xgemm.cc, line 156 \[r1\]](https://reviewable.io:443/reviews/clmathlibraries/clblas/235#-KBt0WUQgMs9pDNS1IV3:-KBt0WURiSH7-4APANOQ:-1222386316) ([raw file](https://github.com/clmathlibraries/clblas/blob/2b56167fbc887acaf42aa7707ab8cfcd3a8c468e/src/library/blas/xgemm.cc#L156)):*
Is there a reason you did not declare and initialize kernel_map in the same statement?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clmathlibraries/clblas/235)*

**Comment by pavanky** [comment]
Hmm. This particular one was because of me. Sorry about that. Is this the only cause of concern for thread safety ? I thought I saw some global variables in the code that's been auto-generated that might cause some issues.

File: src/library/blas/xgemm.cc [hunk 1]
```diff
@@ -135,7 +145,16 @@ void makeGemmKernel(
 {
   //TODO: This will need to be converted to thread local when making clBLAS thread safe
```
Comment (pavanky) [inline src/library/blas/xgemm.cc]: This line can probably go away now.

**Comment by guacamoleo** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/library/blas/xgemm.cc, line 152 \[r1\]](https://reviewable.io:443/reviews/clmathlibraries/clblas/235#-KBszTX9eCAtl7LRXCYC:-KBtEvWZtiziePCZ7_UN:-467915377) ([raw file](https://github.com/clmathlibraries/clblas/blob/2b56167fbc887acaf42aa7707ab8cfcd3a8c468e/src/library/blas/xgemm.cc#L152)):*
fixed

---

*[src/library/blas/xgemm.cc, line 156 \[r1\]](https://reviewable.io:443/reviews/clmathlibraries/clblas/235#-KBt0WUQgMs9pDNS1IV3:-KBtFUM0TN4F8k0CuaW1:-402279040) ([raw file](https://github.com/clmathlibraries/clblas/blob/2b56167fbc887acaf42aa7707ab8cfcd3a8c468e/src/library/blas/xgemm.cc#L156)):*
this is a requirement/limitation of thread local storage. Not using a pointer and calling the constructor of kernel_map() isn't allowed. It may be allowed in c++11.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clmathlibraries/clblas/235)*

**Comment by kknox** [comment]
@pavanky 
I'm curious, can you see the 'reviewable' code review?  Do you have permissions to comment and review?

**Comment by guacamoleo** [comment]
I also tested test-functional *THREAD* and it passes too.

**Comment by pavanky** [comment]
FYI C++11 does have `std::thread_local` that allows you to use constructors. The issues however is the availability of this feature. Visual Studio only got it with VS2015. AppleClang does not implement the feature yet.

**Comment by guacamoleo** [comment]
Yes, we plan to use such c++11 features in future libraries; but for the current clBLAS, we’re trying to stay compatible with older versions.

David E. Tanner
______________________________________________________________________________
Sr Software Engineer  |   Radeon Technologies Group – Open Compute

**Comment by kknox** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/clmathlibraries/clblas/235)*

**Pull request merged**